### PR TITLE
DDFLSBP-150 - Refactored the interest periods configuration. Added ne…

### PR DESCRIPTION
**Pull request by Dresse**
…w default interest period field. Added validating of the new fields.

DON'T MERGE - Dependent on other PR in dpl-react (https://reload.atlassian.net/browse/DDFSOEG-586) 
#### Link to issue

https://reload.atlassian.net/browse/DDFLSBP-150

#### Description

This PR includes:

- Refactoring the old "interest_periods_config" field. This field no longer requires JSON input, but should instead comform to the format [days]-[label] (see screenshot for example). Added description help text.

- Added new field "default_interest_periods_config". The field options of this field is based on the values provided in the "interest_periods_config" field. Added description help text.

- Added validating of the 2 new fields as well as helpful error messages, preventing wrong formats to be added.

- Moved some logic from the "GeneralSettingsForm" into new class "GeneralSettings". This takes care of setting constants as well as having functions for formatting the inputtet values in "interest_periods_config" and "default_interest_period_config" fields. This also includes a function for combining the two configurations into a new settings object which are JSON encoded.

#### Screenshot of the result

The 2 new fields with content:
Screenshot 2023-11-16 at 03 30 33

Error message when choosing invalid default interest period:
Screenshot 2023-11-16 at 03 32 28

Error message when trying to add a new interest period which do not match the valid format:
Screenshot 2023-11-16 at 03 34 03

#### Additional comments or questions

The new JSON encoded interest-periods-config will have the format:
`{
   "interestPeriods":[
      {
         "value":14,
         "label":"14 days"
      },
      {
         "value":30,
         "label":"1 month"
      },
      {
         "value":60,
         "label":"2 months"
      },
      {
         "value":90,
         "label":"3 months"
      },
      {
         "value":180,
         "label":"6 months"
      },
      {
         "value":365,
         "label":"1 year"
      }
   ],
   "defaultInterestPeriod":{
      "value":"14",
      "label":"14 days"
   }
}`

This PR will probably need a sister PR in dpl-react, for using the new config. 

DON'T MERGE - Dependent on other PR in dpl-react (https://reload.atlassian.net/browse/DDFSOEG-586)

### CODE REVIEW

File: web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php [hunk 1]
```diff
@@ -247,9 +252,37 @@ public function buildForm(array $form, FormStateInterface $form_state): array {
...
+   *
+   * @throws \Safe\Exceptions\PcreException
    */
   public function validateForm(array &$form, FormStateInterface $form_state): void {
+    $interestPeriods = $form_state->getValue('interest_periods_config');
+    $defaultInterestPeriod = $form_state->getValue('default_interest_period_config');
+
+    $interestPeriodsArray = explode(PHP_EOL, $interestPeriods);
+    foreach ($interestPeriodsArray as $period) {
+      if (!preg_match('/^[\d]+-[\w ]+$/m', trim($period))) {
```
Comment (spaceo) [inline web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php:272]: Don't know if it makes a difference, but is it intentionally that you wrote a whitespace after the word boundary character class?
`/^[\d]+-[\w ]+$/m`

File: web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php [hunk 2]
```diff
@@ -247,9 +252,37 @@ public function buildForm(array $form, FormStateInterface $form_state): array {
...
+   *
    * {@inheritdoc}
+   *
+   * @throws \Safe\Exceptions\PcreException
    */
   public function validateForm(array &$form, FormStateInterface $form_state): void {
+    $interestPeriods = $form_state->getValue('interest_periods_config');
+    $defaultInterestPeriod = $form_state->getValue('default_interest_period_config');
+
+    $interestPeriodsArray = explode(PHP_EOL, $interestPeriods);
```
Comment (spaceo) [inline web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php:270]: According to drupal coding style variabels should use underscore.
I am not fond of using the data type as a part of the variable name. It should be implicit. So I don't think it is needed to state that it is an array in: `$interestPeriodsArray`

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ */
+class GeneralSettings extends DplReactConfigBase {
+
+  const EXPIRATION_WARNING_DAYS_BEFORE_CONFIG = 6;
+  const RESERVATION_DETAIL_ALLOW_REMOVE_READY_RESERVATIONS = FALSE;
+  const INTEREST_PERIODS_CONFIG = '180-6 months';
+  const DEFAULT_INTEREST_PERIOD_CONFIG = [
+    "value" => "180",
+    "label" => "6 months",
+  ];
```
Comment (spaceo) [inline web/modules/custom/dpl_library_agency/src/GeneralSettings.php:18]: If you change the language to danish defaults it should be corrected here as well.

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      $interestPeriodsConfig['interestPeriods'][] = [
+        'value' => $key,
+        'label' => $value,
+      ];
+    }
+
+    $interestPeriodsConfig['defaultInterestPeriod'] = $defaultInterestPeriod;
+
+    return $interestPeriodsConfig;
+  }
```
Comment (spaceo) [inline web/modules/custom/dpl_library_agency/src/GeneralSettings.php:61]: Throughout this file you use camelCase for variable names. According to drupal coding style they should be lowercase. Isn't phpcs picking this up?

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  /**
+   * Gets interest periods as an array.
+   *
+   * @return mixed[]
+   *   The interest period array.
+   */
+  public function getInterestPeriodsAsArray(): array {
```
Comment (spaceo) [inline web/modules/custom/dpl_library_agency/src/GeneralSettings.php:69]: I dont think it is necessary to specify that the output of the function comes as an array

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    $interestPeriods = $this->loadConfig()->get('interest_periods_config');
+
+    $interestPeriodsArray = [];
+    $optionsArray = explode(PHP_EOL, $interestPeriods);
+    foreach ($optionsArray as $option) {
+      list($days, $label) = explode('-', $option);
+      $interestPeriodsArray[trim($days)] = trim($label);
+    }
```
Comment (spaceo) [inline web/modules/custom/dpl_library_agency/src/GeneralSettings.php:77]: You are using similar functionality [here](https://github.com/danskernesdigitalebibliotek/dpl-cms/pull/475/files#diff-af58b2b66664a675852f35fc9e0ce8987217b12a8e9f7210dc19d070346d78dbR277-R278) maybe you could make a small public method that does that and use that both places so you do not have to repeat it?

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  /**
+   * Gets the default interest period as an array.
+   *
+   * @param array[] $interestPeriods
+   *   The interestPeriods as an array.
+   *
+   * @return array[]
+   *   The default interest period array.
+   */
+  protected function getDefaultInterestPeriodAsArray(array $interestPeriods): array {
```
Comment (spaceo) [inline web/modules/custom/dpl_library_agency/src/GeneralSettings.php:90]: I dont think you need to use the return type as a part of the method name

File: web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php [hunk 8]
```diff
@@ -136,20 +130,30 @@ public function buildForm(array $form, FormStateInterface $form_state): array {
...
+      '#default_value' => $config->get('interest_periods_config') ?? GeneralSettings::INTEREST_PERIODS_CONFIG,
+      '#description' => $this->t('Set the interest periods. The format should be [days]-[label]. New periods will first be available in "Default interest period for reservation" when the form has been saved.', [], ['context' => 'Library Agency Configuration']),
+    ];
+
+    $form['reservations']['default_interest_period_config'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Default interest period for reservation', [], ['context' => 'Library Agency Configuration']),
+      '#options' => $this->generalSettings->getInterestPeriodsAsArray(),
+      '#default_value' => $config->get('default_interest_period_config') ?? GeneralSettings::DEFAULT_INTEREST_PERIOD_CONFIG,
+      '#description' => $this->t('Set the default interest period for reservations.', [], ['context' => 'Library Agency Configuration']),
```
Comment (spaceo) [inline web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php:150]: I don't think it is the best solution with one field being depending on that you have submitted another. It is a confusing user experience, I think.
But I underrstand why you did it, and it is a viable MVP for now, I think. If the customer wants changes here they must telle us. I think we should telle them that we chose to go this way in our next demo

**Review by spaceo — CHANGES_REQUESTED** [review]
Nice work :)
I have some comments though...

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php (hunk 6 above)
Comment (Dresse) [reply web/modules/custom/dpl_library_agency/src/GeneralSettings.php:77]: Yes I am doing something similar
So you were thinking about moving the

File: web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php (hunk 8 above)
Comment (Dresse) [reply web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php:150]: I agree that this is not the best user experience. I tried to do some ajax magic, but had some problems with it. Decided it was not worth the time as I have already used enough.

File: web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php (hunk 1 above)
Comment (Dresse) [reply web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php]: This is on purpose. If not, you are not able to do whitespaces. Example "1 month".

File: web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php (hunk 2 above)
Comment (Dresse) [reply web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php]: Updated this. Please see latest commit.

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php (hunk 4 above)
Comment (Dresse) [reply web/modules/custom/dpl_library_agency/src/GeneralSettings.php]: Updated this. Please see latest commit.

You are right, this is my fault. 
Phpcs is not picking this up on my local dev. I have to look into this.

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php (hunk 5 above)
Comment (Dresse) [reply web/modules/custom/dpl_library_agency/src/GeneralSettings.php]: Updated this. Please see latest commit.

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php (hunk 7 above)
Comment (Dresse) [reply web/modules/custom/dpl_library_agency/src/GeneralSettings.php]: Updated this. Please see latest commit.

File: config/sync/dpl_library_agency.general_settings.yml [hunk 9]
```diff
@@ -1,25 +1,10 @@
...
-    },
-    {
-        "value": "60",
-        "label": "2 months"
-    },
-    {
-        "value": "90",
-        "label": "3 months"
-    },
+interest_periods_config: '180-6 months'
```
[the remark this answers is not in the document]
Comment (Dresse) [reply config/sync/dpl_library_agency.general_settings.yml]: Updated this. Please see latest commit.

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php (hunk 5 above)
Comment (spaceo) [reply web/modules/custom/dpl_library_agency/src/GeneralSettings.php]: I was refering to the name of the function :)

**Review by spaceo — APPROVED** [review]
One thing:
This should not be a part of the PR, I think:
web/profiles/dpl_cms/translations/da.po

Otherwise LGTM 👍

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php (hunk 4 above)
Comment (spaceo) [reply web/modules/custom/dpl_library_agency/src/GeneralSettings.php]: 👍

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php (hunk 6 above)
Comment (spaceo) [reply web/modules/custom/dpl_library_agency/src/GeneralSettings.php]: Yes could be a static method on GeneralSettings

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php (hunk 7 above)
Comment (spaceo) [reply web/modules/custom/dpl_library_agency/src/GeneralSettings.php]: 👍

File: web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php (hunk 8 above)
Comment (spaceo) [reply web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php]: 👍

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php (hunk 3 above)
Comment (spaceo) [reply web/modules/custom/dpl_library_agency/src/GeneralSettings.php]: 👍

File: web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php (hunk 2 above)
Comment (spaceo) [reply web/modules/custom/dpl_library_agency/src/Form/GeneralSettingsForm.php]: 👍

File: web/modules/custom/dpl_library_agency/src/GeneralSettings.php (hunk 6 above)
Comment (Dresse) [reply web/modules/custom/dpl_library_agency/src/GeneralSettings.php]: I have done this now in my latest commit and removed translations/da.po from the PR altogether.

**Review by spaceo — APPROVED** [review]
LGTM 👍

**Pull request merged**